google-cloud-storage/src/test/java/com/google/cloud/storage/conformance/retry/RpcMethodMappings.java

```diff
@@ -0,0 +1,1942 @@
+    Objects.get(a);
+    Objects.insert(a);
+    Objects.list(a);
+    Objects.patch(a);
+    Objects.update(a);
+    Objects.compose(a);
+    Objects.rewrite(a);
+    Objects.copy(a);
+
+    ServiceAccount.get(a);
+    ServiceAccount.put(a);
+
+    validateMappingDefinitions(a);
+
+    funcMap = Multimaps.index(a, RpcMethodMapping::getMethod);
+    reportMappingSummary();
+  }
+
+  public Collection<RpcMethodMapping> get(RpcMethod key) {
+    return funcMap.get(key);
+  }
+
+  public Set<Integer> differenceMappingIds(Set<Integer> usedMappingIds) {
+    return Sets.difference(
+        funcMap.values().stream().map(RpcMethodMapping::getMappingId).collect(Collectors.toSet()),
+        usedMappingIds);
+  }
+
+  private void validateMappingDefinitions(ArrayList<RpcMethodMapping> a) {
+    ListMultimap<Integer, RpcMethodMapping> idMappings =
+        MultimapBuilder.hashKeys()
+            .arrayListValues()
+            .build(Multimaps.index(a, RpcMethodMapping::getMappingId));
+    String duplicateIds =
+        idMappings.asMap().entrySet().stream()
+            .filter(e -> e.getValue().size() > 1)
+            .map(Entry::getKey)
+            .map(i -> Integer.toString(i))
+            .collect(Collectors.joining(", "));
+    if (!duplicateIds.isEmpty()) {
+      String message = "duplicate mapping ids present: [" + duplicateIds + "]";
+      throw new IllegalStateException(message);
+    }
+  }
+
+  private void reportMappingSummary() {
+    int mappingCount = funcMap.values().stream().mapToInt(m -> 1).sum();
+    LOGGER.info("Current total number of mappings defined: " + mappingCount);
+    String counts =
+        funcMap.asMap().entrySet().stream()
+            .map(
+                e -> {
+                  RpcMethod rpcMethod = e.getKey();
+                  Collection<RpcMethodMapping> mappings = e.getValue();
+                  return String.format(
+                      "\t%s.%s: %d",
+                      rpcMethod
+                          .getClass()
+                          .getName()
+                          .replace("com.google.cloud.storage.conformance.retry.RpcMethod$", "")
+                          .replace("$", "."),
+                      rpcMethod,
+                      mappings.size());
+                })
+            .sorted()
+            .collect(Collectors.joining("\n", "\n", ""));
+    LOGGER.info("Current number of mappings per rpc method: " + counts);
+    OptionalInt max =
+        funcMap.values().stream().map(RpcMethodMapping::getMappingId).mapToInt(i -> i).max();
+    if (max.isPresent()) {
+      LOGGER.info(String.format("Current max mapping index is: %d%n", max.getAsInt()));
+    } else {
+      throw new IllegalStateException("No mappings defined");
+    }
+  }
+
+  static final class Mappings {
+
+    static final class BucketAcl {
+
+      private static void delete(ArrayList<RpcMethodMapping> a) {
+        a.add(
+            RpcMethodMapping.newBuilder(1, bucket_acl.delete)
```

Comment:
How do we not use constants as an ID? I suspect we will get burned easily unless I'm missing something.

Reply:
the id's don't need to be stable themselves, only unique. If any id is determined to be non-unique an error will be thrown and the suite will fail, for example:
```
java.lang.IllegalStateException: duplicate mapping ids present: [2, 5, 6]
	at com.google.cloud.storage.conformance.retry.RpcMethodMappings.validateMappingDefinitions(RpcMethodMappings.java:195)
	at com.google.cloud.storage.conformance.retry.RpcMethodMappings.<init>(RpcMethodMappings.java:166)
```

Additonaly, any ids which aren't used during test case resolution will be reported as well
```
Sep 15, 2021 1:21:46 PM com.google.cloud.storage.conformance.retry.ITRetryConformanceTest$RetryTestCaseResolver validateGeneratedTestCases
WARNING: Declared but unused mappings with ids: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
```

If we were to want named constants, we'd need to define (so far) 123 of them.

Reply:
Thanks for the additional context, would we lose anything if you used a static counter which increments on each instant of a mapping instead?

Reply:
With JUnit 4 we don't easily have a way of pointing to the mapping as the actual test definition, so the static id serves as a ctrl+f way of finding a test fairly easily. If we had access to JUnit 5 there are supposed to be ways which we could "point" to the mapping.